### PR TITLE
`Programming exercises`: Export files embedded with HTML syntax

### PR DESCRIPTION
**Pull request by tobias-lippert**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The export of a programming exercise doesn't contain files embedded using HTML syntax because the regex doesn't match them.

### Description

Added a second regex to also match files embedded with HTML syntax refactored some code to reuse logic for both regex expressions.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming exercise

1. Embedded some files in the problem statement using both markdown and html syntax
2. Export (Download) the exercise
3. Verify that the export contains all embedded files.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |Comment|
|------------|--------------:|---------------------:|--------------------:|
| [ProgrammingExerciseExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5418/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseExportService.html) | 86% | ✅ | my changes are covered

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 1]
```diff
@@ -10,7 +10,14 @@
...
 import java.time.format.DateTimeFormatter;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:20]: Is this change intentional?

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 2]
```diff
@@ -90,7 +103,9 @@ public class ProgrammingExerciseExportService {
 
     public static final String EXPORTED_EXERCISE_PROBLEM_STATEMENT_FILE_PREFIX = "Problem-Statement";
 
-    private static final String EMBEDDED_FILE_REGEX = "\\[.*] *\\(/api/files/markdown/.*\\)";
+    private static final String EMBEDDED_FILE_MARKDOWN_SYNTAX_REGEX = "\\[.*] *\\(/api/files/markdown/.*\\)";
+
+    private static final String EMBEDDED_FILE_HTML_SYNTAX_REGEX = "<img src=\"/api/files/markdown/.*?\" .*?>";
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:108]: Why do you need the question marks after the `.*`?

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 3]
```diff
@@ -170,13 +185,63 @@ private void exportProblemStatementAndEmbeddedFiles(ProgrammingExercise exercise
...
+                Files.copy(imageFilePath, imageExportPath);
+            }
+            catch (IOException e) {
+                exportErrors.add("Failed to copy embedded files: " + e.getMessage());
+                log.warn("Could not copy embedded file {} for exercise with id {}", fileName, exercise.getId());
+            }
+        }
+    }
+
+    private void copyFilesEmbeddedWithMarkdownSyntax(ProgrammingExercise exercise, List<String> exportErrors, Set<String> embeddedFilesWithMarkdownSyntax, Path embeddedFilesDir) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:229]: Can you swap this method with the one above? I think that order would make more sense

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java (hunk 1 above)
Comment (tobias-lippert) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:20]: no

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me now

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 4]
```diff
@@ -168,15 +179,100 @@ private void exportProblemStatementAndEmbeddedFiles(ProgrammingExercise exercise
...
+            // avoid matching other closing ] or () in the squared brackets by getting the index of the last ]
+            String lastPartOfMatchedString = embeddedFile.substring(embeddedFile.lastIndexOf("]") + 1);
+            String filePath = lastPartOfMatchedString.substring(lastPartOfMatchedString.indexOf("(") + 1, lastPartOfMatchedString.indexOf(")"));
+            constructFilenameAndCopyFile(exercise, exportErrors, embeddedFilesDir, filePath);
+        }
+    }
+
+    /**
+     * Copies the files that are embedded with html syntax to the embedded files' directory.
+     * <p>
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:219]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 5]
```diff
@@ -168,15 +179,100 @@ private void exportProblemStatementAndEmbeddedFiles(ProgrammingExercise exercise
...
+            int indexOfFirstQuotationMark = embeddedFile.indexOf('"');
+            String filePath = embeddedFile.substring(embeddedFile.indexOf("src=") + 5, embeddedFile.indexOf('"', indexOfFirstQuotationMark + 1));
+            constructFilenameAndCopyFile(exercise, exportErrors, embeddedFilesDir, filePath);
+        }
+    }
 
-        Matcher matcher = Pattern.compile(EMBEDDED_FILE_REGEX).matcher(exercise.getProblemStatement());
+    /**
+     * Extracts the filename from the matched string and copies the file to the embedded files' directory.
+     * <p>
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:236]:
```suggestion
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java [hunk 6]
```diff
@@ -168,15 +179,100 @@ private void exportProblemStatementAndEmbeddedFiles(ProgrammingExercise exercise
...
+            catch (IOException e) {
+                exportErrors.add("Failed to copy embedded files: " + e.getMessage());
+                log.warn("Could not copy embedded file {} for exercise with id {}", fileName, exercise.getId());
+            }
+        }
+    }
+
+    /**
+     * Checks for matches in the problem statement and creates a directory for the embedded files.
+     * <p>
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseExportService.java:261]:
```suggestion
```

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM
Tested on TS2, works

**Review by JohannesStoehr — APPROVED** [review]

**Review by dearjasmina — APPROVED** [review]
tested in the testing session, works as described

**Review by laurenzfb — APPROVED** [review]
Tested in testing session. Worked as expected!

**Review by Strohgelaender — APPROVED** [review]
tested in testing session, works as expected.

**Pull request merged**